### ISSUE TITLE
Question about `--alter-wrapper` and wait_timeout

### ISSUE THREAD

**Issue by tomkrouper** [skeema/skeema#43]
We have yet to run a `skeema push` in production in an automated way because I felt uneasy about the output of the `skeema diff` command. I've created an arbitrary example to illustrate my concern: 
```
mysql> SELECT @@wait_timeout;
+----------------+
| @@wait_timeout |
+----------------+
|          28800 |
+----------------+
1 row in set (0.00 sec)

mysql> set @@wait_timeout = 5;
Query OK, 0 rows affected (0.00 sec)

mysql> \! sleep 10
mysql> SELECT @@wait_timeout;
ERROR 2006 (HY000): MySQL server has gone away
No connection. Trying to reconnect...
Connection id:    290
Current database: *** NONE ***

+----------------+
| @@wait_timeout |
+----------------+
|          28800 |
+----------------+
1 row in set (0.00 sec)
```

What I'm trying to show here is if we had a script in `--alter-wrapper` that we were to run `gh-ost` or `pt-online-schema-change` that ran longer than `wait_timeout`, the MySQL connection would be lost. I'm not sure what that means for the script that is running. Will `skeema` still be able to get back the output from the `--alter-wrapper` script?

**Comment by evanelias** [comment]
Excellent question! Using a low wait_timeout can work fine, but currently it requires adding something to your skeema configuration. I'll improve this soon so it won't need configuration.

With an aggressive (under 30 second) server global `wait_timeout`, currently you'd need to configure Skeema to specifically use a session-level wait_timeout using its [connect-options](https://github.com/skeema/skeema/blob/master/doc/options.md#connect-options) flag. For example, on the command-line:
```skeema diff --connect-options='wait_timeout=60'```

Or more ideally, in a `.skeema` option file at the root of the schema repo:
```connect-options='wait_timeout=60'```

This does two things:

1. All connections in skeema's connection pool will automatically set a session-level `wait_timeout` upon starting the connection. (You can configure any other MySQL session variable in this way too)

2. As a special case for wait_timeout, Skeema's tengo package will [see the value and configure its connection pool's own client-side idle timeout accordingly](https://github.com/skeema/tengo/blob/master/instance.go#L140). This way, the connection pool will properly evict idle connections *before* the server kills them, which avoids nasty problems in the golang mysql driver (see lengthy driver issue https://github.com/go-sql-driver/mysql/issues/657).

I'm just using 60 here as an example. Even if you set it to the exact same value as the global server default, this setting is still beneficial. Tengo only configures the client-side timeout correctly if you explicitly provide a wait_timeout (e.g. via skeema's `connect-options`). Without an explicit `wait_timeout`, it isn't smart enough to otherwise proactively query the server to determine the global default, but I'll improve this soon.

Regarding "uneasy about the output of the skeema diff command", are you referring to how `skeema diff`'s output includes the `\!` prefix for any shelled-out external wrapper command? I should document this better, but it just does this for display purposes. The idea is that all STDOUT of `skeema diff` could, if desired, be manually copied-and-pasted into the `mysql` command-line client, if the user wanted to. However, `skeema push` doesn't actually use or rely on the `mysql` command-line client itself when performing its operations.

When running an `alter-wrapper` script, `skeema push` just redirects the new process's STDOUT and STDERR to its own, rather than capturing/buffering the output. It only cares about the process's exit code (zero vs non-zero) to tell if the script was successful. 

If `skeema` encounters a connection timeout while the `alter-wrapper` script is running, this should be fine, as long as `connect-options` is configured as above. Skeema will proactively evict its own idle sessions and lazily reconnect when needed.

**Comment by tomkrouper** [comment]
Awesome. Thanks for the detailed response. Great information. Gonna go ahead an close this since my question was answered.

**Comment by evanelias** [comment]
Just released [Skeema v1.0.6](https://github.com/skeema/skeema/releases/tag/v1.0.6), which has the wait_timeout improvements mentioned earlier in this issue. It's no longer necessary to use `--connect-options` to explicitly specify a session-level wait_timeout; Skeema will now always auto-detect the server default, and will ensure its connection pool cleans up old connections before they hit the wait_timeout.
